### ISSUE TITLE
How to load many items with DataLoader?

### ISSUE THREAD

**Issue by acelot** [graphql-dotnet/graphql-dotnet#1153]
I have the following scheme:
```gql
type Query {
  articles: [Article!]!
}

type Article {
  id: ID!,
  title: String!,
  authors: [Author!]!
}

type Author {
  id: ID!,
  name: String!
}
```

Typical N+1 problem when you need to load authors for each article. Authors are stored in `articles` table as an array of integers in `authors` column in PostgreSQL.

Articles table example:

|id|title|authors|
|--|----|---------|
|`1`|Article 1|`[1,2,3]`|
|`2`|Article 2|`[1,2]`|
|`3`|Article 3|`[3,1]`|

What I need to get from graphql:

```json
{
  "data": {
    "articles": [
      {
        "id": 1,
        "title": "Article 1",
        "authors": [
          {
            "id": 1,
            "name": "John Snow"
          },
          {
            "id": 2,
            "name": "Daenerys Targaryen"
          },
          {
            "id": 3,
            "name": "Tyrion Lannister"
          }
        ]
      },
      {
        "id": 2,
        "title": "Article 2",
        "authors": [
          {
            "id": 1,
            "name": "John Snow"
          },
          {
            "id": 2,
            "name": "Daenerys Targaryen"
          }
        ]
      },
      {
        "id": 3,
        "title": "Article 3",
        "authors": [
          {
            "id": 3,
            "name": "Tyrion Lannister"
          },
          {
            "id": 1,
            "name": "John Snow"
          }
        ]
      }
    ]
  }
}
```

It is expected that a single request must be executed to get all the authors.
What I am trying to do:
```cs
public class ArticleType : ObjectGraphType<Article>
{
    private NpgsqlConnection conn;

    public ArticleType(IDataLoaderContextAccessor accessor)
    {
        Field(a => a.Id).Name("id");

        Field(a => a.Title).Name("title");

        Field<ListGraphType<AuthorType>>().Name("authors")
            .ResolveAsync(async (ctx) =>
            {
                var authorsLoader = accessor.Context.GetOrAddCollectionBatchLoader<int, Author>("getAuthorsByIds", GetAuthorsByIds);
                int[] authorsIds = ctx.Source.AuthorsIds;
                var authors = await authorsLoader.LoadAsync(authorsIds);
                ...
            });
    }

    public Task<ILookup<int, Author>> GetAuthorsByIds(IEnumerable<int> ids)
    {
        ...
    }
}
```

But unfortunately the `LoadAsync()` method takes only single key as parameter. How to load many keys simultaneously? Or should I get them one by one with awaits?

**Comment by dBug404** [comment]
Maybe this would be possible somehow, but i think its not supposed to work like this.

Usually you have an id and for that id you get an collection of items.

So in your example you could use the articleId instead of the List of authorIds.



`   
 public ArticleType(IDataLoaderContextAccessor accessor)
    {
        Field(a => a.Id).Name("id");

        Field(a => a.Title).Name("title");

        Field<ListGraphType<AuthorType>>().Name("authors")
            .ResolveAsync(async (ctx) =>
            {
                var authorsLoader = accessor.Context.GetOrAddCollectionBatchLoader<int, Author>("getAuthorsByIds", GetAuthorsByIds);
                int articleId = ctx.Source.Id;
                var authors = await authorsLoader.LoadAsync(articleId);
                ...
            });
    }

    public Task<ILookup<int, Author>> GetAuthorsByArticleId(int articleId)
    {
        ...
    }
}`

**Comment by acelot** [comment]
Thanks for advice, I can do like that. But this method seems to me not optimal, because I already have authors IDs in the context and it would be easier just to get authors on them. And according to this method will have to apply join like this: `SELECT unnest("authors") LEFT JOIN authors`.

**Comment by johnrutherford** [comment]
@acelot, We don't currently have support for what you're trying to do. The DataLoader expects you to pass in a single key to each call. 

We could probably fairly easily add a method like this to the batch loader to pass in multiple keys:

``` csharp
public Task<T[]> LoadAsync(IEnumerable<TKey> keys)
```

For now you could try doing this:

``` csharp
var authorsLoader = accessor.Context.GetOrAddBatchLoader<int, Author>("getAuthorsByIds", GetAuthorsByIds);
int[] authorsIds = ctx.Source.AuthorsIds;

var tasks = new List<Task<Author>>(authorsIds.Length);

foreach (int authorId in authorsIds)
{
    tasks.Add(authorsLoader.LoadAsync(authorId));
}

var authors = await Task.WhenAll(tasks);
```

Note that this uses the batch data loader, not the collection batch data loader. And it's also very important to await all of the tasks at once or it will deadlock. This way should work, but I can't guarantee it.

**Comment by benmccallum** [comment]
I've currently got a need for this and was thinking I'd PR it in after checking the issues. Will give it a go.

**Comment by benmccallum** [comment]
PR is up with tests. In the meantime, here's an extension method that I'm about to test locally now that you can drop into your project until a v3 is released.

```csharp
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;
using GraphQL.DataLoader;

namespace GraphQLService
{
    public static class IDataLoaderExtensions
    {
        public static async Task<T[]> LoadAsync<TKey, T>(this IDataLoader<TKey, T> dataLoader, IEnumerable<TKey> keys)
        {
            var tasks = new List<Task<T>>(keys.Count());

            foreach (var key in keys)
            {
                tasks.Add(dataLoader.LoadAsync(key));
            }

            return await Task.WhenAll(tasks);
        }
    }
}
```

**Comment by sungam3r** [comment]
It’s better to make these methods extension method. In fact, they do not carry additional logic inside the dataloder itself. They just iterate an already existing method with one key. Therefore, expanding the `IDataLoader` interface seems redundant.

**Comment by benmccallum** [comment]
Because it's more performant since it's be a static method? 

Adding it to the interface was necessary as the methods to get a dataloader actually only return the interface. 

I did feel like adding the changes to the tests were kinda weird a d that I could just make them pure unit tests that returned the list of tasks.

Can certainly change it all to an extension method and write a test to that if everyone agrees?

**Comment by sungam3r** [comment]
> Because it's more performant since it's be a static method?

No. Because it is not a logic inside dataloader but around it existing method.

> Adding it to the interface was necessary as the methods to get a dataloader actually only return the interface.

I don't understand your motivation. User will call your extension method on returned interface.

**Comment by benmccallum** [comment]
> No. Because it is not a logic inside dataloader but around it existing method.
What ever did people did before extension methods... 😅

I meant that adding it to the interface was necessary in my current implementation as an instance method. Obviously my above snippet is against the interface and will be fine when it's an extension method.

**Comment by sungam3r** [comment]
> What ever did people did before extension methods...

Well, if you extend interface - it's a breaking change. If you **can** add extension method **instead** - you are lucky. 

> Obviously my above snippet is against the interface and will be fine when it's an extension method.

Ok.

**Comment by benmccallum** [comment]
You raise a good point about it being a breaking change.

**Comment by benmccallum** [comment]
Have updated to use extension method and undone all my adjustments to previous tests in favour of just testing the extension method.

**Comment by sungam3r** [comment]
Ok.

> You raise a good point about it being a breaking change.

The long development of libraries and their stacks affects.

**Comment by benmccallum** [comment]
I guess another way would be a default interface member implementation in C# 8, but extension method would always be preferred unless we had people writing their own dataloader and wanting the flexibility to modify behaviour of this.

**Comment by sungam3r** [comment]
Default interface members is the feature that makes me tremble )

**Comment by sungam3r** [comment]
Closed thanks to #1282
